banyand/metadata/schema/utils.go

```diff
@@ -0,0 +1,140 @@
+					cmp.Equal(aStream.GetEntity().GetTagNames(), bStream.GetEntity().GetTagNames()) && // entities
+					tagFamilySpecsEqual(aStream.GetTagFamilies(), bStream.GetTagFamilies()) // tagFamilies
+			}
+		}
+
+		return false
+	}
+
+	measureEqualityChecker = func(a, b proto.Message) bool {
+		if bMeasure, ok := b.(*databasev1.Measure); ok {
+			if aMeasure, ok := a.(*databasev1.Measure); ok {
+				return metadataEqual(aMeasure.GetMetadata(), bMeasure.GetMetadata()) &&
+					resourceOptEqual(aMeasure.GetOpts(), bMeasure.GetOpts()) && // resourceOpts
+					cmp.Equal(aMeasure.GetEntity().GetTagNames(), bMeasure.GetEntity().GetTagNames()) && // entities
+					tagFamilySpecsEqual(aMeasure.GetTagFamilies(), bMeasure.GetTagFamilies()) && // tagFamilies
+					intervalSpecsEqual(aMeasure.GetIntervalRules(), bMeasure.GetIntervalRules()) && // intervalRules
+					fieldSpecsEqual(aMeasure.GetFields(), bMeasure.GetFields()) // fields
+			}
+		}
+
+		return false
+	}
+
+	indexRuleChecker = func(a, b proto.Message) bool {
+		if bIndexRule, ok := b.(*databasev1.IndexRule); ok {
+			if aIndexRule, ok := a.(*databasev1.IndexRule); ok {
+				return metadataEqual(aIndexRule.GetMetadata(), bIndexRule.GetMetadata()) &&
+					cmp.Equal(aIndexRule.GetTags(), bIndexRule.GetTags()) && // tags
+					aIndexRule.GetType() == bIndexRule.GetType() &&
+					aIndexRule.GetLocation() == bIndexRule.GetLocation()
+			}
+		}
+
+		return false
+	}
+
+	indexRuleBindingChecker = func(a, b proto.Message) bool {
+		if bIndexRuleBinding, ok := b.(*databasev1.IndexRuleBinding); ok {
+			if aIndexRuleBinding, ok := a.(*databasev1.IndexRuleBinding); ok {
+				return metadataEqual(aIndexRuleBinding.GetMetadata(), bIndexRuleBinding.GetMetadata()) &&
+					cmp.Equal(aIndexRuleBinding.GetRules(), bIndexRuleBinding.GetRules()) && // rules
+					timestamppbEqual(aIndexRuleBinding.GetBeginAt(), bIndexRuleBinding.GetBeginAt()) &&
+					timestamppbEqual(aIndexRuleBinding.GetExpireAt(), bIndexRuleBinding.GetExpireAt()) &&
+					subjectEqual(aIndexRuleBinding.GetSubject(), bIndexRuleBinding.GetSubject())
+			}
+		}
+
+		return false
+	}
+)
+
+var (
+	checkerMap = map[Kind]equalityChecker{
+		KindIndexRuleBinding: indexRuleBindingChecker,
+		KindIndexRule:        indexRuleChecker,
+		KindMeasure:          measureEqualityChecker,
+		KindStream:           streamEqualityChecker,
+	}
+)
+
+func metadataEqual(a, b *commonv1.Metadata) bool {
+	return a.GetName() == b.GetName() && a.GetGroup() == b.GetGroup()
+}
+
+func timestamppbEqual(a, b *timestamppb.Timestamp) bool {
+	return a.GetSeconds() == b.GetSeconds() && a.GetNanos() == b.GetNanos()
+}
+
+func subjectEqual(a, b *databasev1.Subject) bool {
+	return a.GetCatalog() == b.GetCatalog() && a.GetName() == b.GetName()
+}
+
+func resourceOptEqual(a, b *databasev1.ResourceOpts) bool {
+	return a.GetShardNum() == b.GetShardNum() && // opts
+		a.GetTtl().GetVal() == b.GetTtl().GetVal() &&
```

Comment:
Could `a.GetTtl()` return `nil`?

Reply:
I've migrated to `protocmp`, https://pkg.go.dev/google.golang.org/protobuf/testing/protocmp.

I suppose this is the much better way to do equality check.